File: ops/worker/metadata_extractor.py
"""
Metadata extraction and file validation
Uses mutagen for audio file metadata
"""
import os
from pathlib import Path
from typing import Optional, Dict, Any
from dataclasses import dataclass
from mutagen import File as MutagenFile
from mutagen.easyid3 import EasyID3
from mutagen.mp3 import MP3
from mutagen.flac import FLAC
from mutagen.mp4 import MP4
from mutagen.oggvorbis import OggVorbis


@dataclass
class AudioMetadata:
    """Extracted audio metadata"""
    artist: Optional[str] = None
    album: Optional[str] = None
    title: Optional[str] = None
    track_number: Optional[int] = None
    year: Optional[int] = None
    genre: Optional[str] = None
    duration: Optional[float] = None  # seconds
    bitrate: Optional[int] = None  # kbps
    sample_rate: Optional[int] = None  # Hz
    channels: Optional[int] = None
    codec: Optional[str] = None
    file_size: int = 0

    def is_valid(self) -> bool:
        """Check if metadata contains minimum required fields"""
        return bool(self.artist and self.title)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            "artist": self.artist,
            "album": self.album,
            "title": self.title,
            "track_number": self.track_number,
            "year": self.year,
            "genre": self.genre,
            "duration": self.duration,
            "bitrate": self.bitrate,
            "sample_rate": self.sample_rate,
            "channels": self.channels,
            "codec": self.codec,
            "file_size": self.file_size
        }


class MetadataExtractor:
    """Extract metadata from audio files"""

    SUPPORTED_EXTENSIONS = {'.mp3', '.flac', '.m4a', '.ogg', '.opus', '.wma'}

    @staticmethod
    def is_audio_file(file_path: Path) -> bool:
        """Check if file is a supported audio format"""
        return file_path.suffix.lower() in MetadataExtractor.SUPPORTED_EXTENSIONS

    @staticmethod
    def extract(file_path: Path) -> Optional[AudioMetadata]:
        """
        Extract metadata from audio file

        Args:
            file_path: Path to audio file

        Returns:
            AudioMetadata object or None if extraction fails
        """
        if not file_path.exists():
            return None

        if not MetadataExtractor.is_audio_file(file_path):
            return None

        try:
            audio = MutagenFile(str(file_path), easy=True)
            if audio is None:
                return None

            metadata = AudioMetadata()
            metadata.file_size = file_path.stat().st_size

            # Extract common tags
            metadata.artist = MetadataExtractor._get_tag(audio, 'artist')
            metadata.album = MetadataExtractor._get_tag(audio, 'album')
            metadata.title = MetadataExtractor._get_tag(audio, 'title')
            metadata.genre = MetadataExtractor._get_tag(audio, 'genre')

            # Track number
            track_str = MetadataExtractor._get_tag(audio, 'tracknumber')
            if track_str:
                try:
                    # Handle "1/12" format
                    metadata.track_number = int(track_str.split('/')[0])
                except (ValueError, IndexError):
                    pass

            # Year
            year_str = MetadataExtractor._get_tag(audio, 'date')
            if year_str:
                try:
                    metadata.year = int(year_str[:4])
                except (ValueError, IndexError):
                    pass

            # Audio properties
            if hasattr(audio.info, 'length'):
                metadata.duration = audio.info.length

            if hasattr(audio.info, 'bitrate'):
                metadata.bitrate = audio.info.bitrate // 1000  # Convert to kbps

            if hasattr(audio.info, 'sample_rate'):
                metadata.sample_rate = audio.info.sample_rate

            if hasattr(audio.info, 'channels'):
                metadata.channels = audio.info.channels

            # Codec detection
            metadata.codec = MetadataExtractor._detect_codec(file_path, audio)

            return metadata

        except Exception as e:
            print(f"Error extracting metadata from {file_path}: {e}")
            return None

    @staticmethod
    def _get_tag(audio, key: str) -> Optional[str]:
        """Get tag value from audio file"""
        if key in audio:
            value = audio[key]
            if isinstance(value, list) and len(value) > 0:
                return str(value[0])
            elif isinstance(value, str):
                return value
        return None

    @staticmethod
    def _detect_codec(file_path: Path, audio) -> Optional[str]:
        """Detect audio codec"""
        ext = file_path.suffix.lower()

        if ext == '.mp3':
            return 'MP3'
        elif ext == '.flac':
            return 'FLAC'
        elif ext in ['.m4a', '.mp4']:
            return 'AAC'
        elif ext in ['.ogg', '.opus']:
            return 'OGG/Opus'
        elif ext == '.wma':
            return 'WMA'

        return None

    @staticmethod
    def normalize_filename(metadata: AudioMetadata, original_path: Path) -> str:
        """
        Generate normalized filename from metadata

        Format: Artist - Title.ext
        Falls back to original name if metadata is insufficient
        """
        if not metadata.is_valid():
            return original_path.name

        artist = MetadataExtractor._sanitize_filename(metadata.artist)
        title = MetadataExtractor._sanitize_filename(metadata.title)
        ext = original_path.suffix

        return f"{artist} - {title}{ext}"

    @staticmethod
    def generate_library_path(metadata: AudioMetadata, library_root: Path) -> Optional[Path]:
        """
        Generate organized library path from metadata

        Structure: library_root/Artist/Album/Track - Title.ext
        Returns None if metadata is insufficient
        """
        if not metadata.is_valid():
            return None

        artist = MetadataExtractor._sanitize_filename(metadata.artist)
        album = MetadataExtractor._sanitize_filename(metadata.album or "Unknown Album")
        title = MetadataExtractor._sanitize_filename(metadata.title)

        # Build filename with track number if available
        if metadata.track_number:
            filename = f"{metadata.track_number:02d} - {title}"
        else:
            filename = title

        return library_root / artist / album / filename

    @staticmethod
    def _sanitize_filename(text: str) -> str:
        """Remove/replace characters not safe for filenames"""
        if not text:
            return "Unknown"

        # Replace problematic characters
        replacements = {
            '/': '-',
            '\\': '-',
            ':': '-',
            '*': '',
            '?': '',
            '"': "'",
            '<': '',
            '>': '',
            '|': '-',
            '\0': ''
        }

        for old, new in replacements.items():
            text = text.replace(old, new)

        # Remove leading/trailing whitespace and dots
        text = text.strip('. ')

        return text or "Unknown"


class FileValidator:
    """Validate audio files"""

    MIN_FILE_SIZE = 100 * 1024  # 100 KB
    MAX_FILE_SIZE = 500 * 1024 * 1024  # 500 MB

    @staticmethod
    def validate(file_path: Path) -> tuple[bool, Optional[str]]:
        """
        Validate audio file

        Args:
            file_path: Path to file

        Returns:
            (is_valid, error_message)
        """
        if not file_path.exists():
            return False, "File does not exist"

        if not file_path.is_file():
            return False, "Not a file"

        # Check file size
        file_size = file_path.stat().st_size
        if file_size < FileValidator.MIN_FILE_SIZE:
            return False, f"File too small ({file_size} bytes)"

        if file_size > FileValidator.MAX_FILE_SIZE:
            return False, f"File too large ({file_size} bytes)"

        # Check if it's an audio file
        if not MetadataExtractor.is_audio_file(file_path):
            return False, f"Unsupported format: {file_path.suffix}"

        # Try to extract metadata
        metadata = MetadataExtractor.extract(file_path)
        if metadata is None:
            return False, "Cannot read audio metadata"

        if not metadata.is_valid():
            return False, "Missing required metadata (artist or title)"

        return True, None

File: backend/internal/services/metadata_extractor.go
package services

import (
	"bytes"
	"crypto/md5"
	"encoding/json"
	"fmt"
	"image"
	_ "image/gif" // register decoders for cover-art round-tripping
	_ "image/jpeg"
	_ "image/png"
	"io"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/bogem/id3v2/v2"
	"github.com/dhowden/tag"
	"github.com/gcottom/audiometa/v3"
	"github.com/go-flac/flacpicture/v2"
	"github.com/go-flac/go-flac/v2"
	_ "golang.org/x/image/webp" // register WebP cover-art decoding
)

type AudioMetadata struct {
	Artist      string
	AlbumArtist string
	Album       string
	Title       string
	TrackNumber int
	Year        int
	Format      string
	FileSize    int64
}

func (m *AudioMetadata) IsValid() bool {
	return m.Artist != "" && m.Title != ""
}

type MetadataExtractor struct{}

func NewMetadataExtractor() *MetadataExtractor {
	return &MetadataExtractor{}
}

// MinimumCoverArtSize is the minimum byte size for a valid cover art image (2KB).
const MinimumCoverArtSize = 2048

// detectImageMimeType detects the MIME type of image data from magic bytes.
func detectImageMimeType(data []byte) string {
	if len(data) < 4 {
		return "image/jpeg" // safe default
	}
	switch {
	case data[0] == 0xFF && data[1] == 0xD8 && data[2] == 0xFF:
		return "image/jpeg"
	case data[0] == 0x89 && data[1] == 0x50 && data[2] == 0x4E && data[3] == 0x47:
		return "image/png"
	case data[0] == 0x47 && data[1] == 0x49 && data[2] == 0x46:
		return "image/gif"
	case len(data) >= 12 && string(data[8:12]) == "WEBP": // WEBP: RIFF....WEBP at offset 8
		return "image/webp"
	default:
		return "image/jpeg" // fallback for unknown formats
	}
}

// EmbedCoverArt embeds image data into the audio file
func (e *MetadataExtractor) EmbedCoverArt(filePath string, artData []byte) error {
	if len(artData) < MinimumCoverArtSize {
		return fmt.Errorf("cover art image too small (%d bytes), likely invalid", len(artData))
	}
	ext := strings.ToLower(filepath.Ext(filePath))
	switch ext {
	case ".mp3":
		return e.embedMP3(filePath, artData)
	case ".flac":
		return e.embedFLAC(filePath, artData)
	case ".m4a", ".ogg":
		return e.embedGeneric(filePath, artData)
	default:
		return fmt.Errorf("unsupported file format for cover art embedding: %s", ext)
	}
}

// embedGeneric uses audiometa v3 to embed cover art into M4A/OGG files.
//
// Upstream note: audiometa v1.3.1 panicked on malformed/undecodable MP4 covr
// atoms (unchecked covr.(*image.Image) assertion); v3 returns errors instead.
// The recover() guard is kept as a backstop for any other latent panic in the
// tagging stack — it should never fire with v3.
func (e *MetadataExtractor) embedGeneric(filePath string, artData []byte) (err error) {
	defer func() {
		if r := recover(); r != nil {
			slog.Warn("audiometa panicked while embedding cover art; skipping embed",
				"path", filePath, "panic", r)
			err = nil
		}
	}()

	// v3's Tag model carries decoded images: decode the art bytes first.
	img, _, err := image.Decode(bytes.NewReader(artData))
	if err != nil {
		return fmt.Errorf("cover art is not a decodable image: %w", err)
	}

	f, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("failed to open file for tagging: %w", err)
	}
	defer f.Close()

	t, err := audiometa.OpenTag(f)
	if err != nil {
		return fmt.Errorf("failed to open tags: %w", err)
	}
	t.SetCoverArt(&img)

	// v3 Save() writes to an io.Writer; M4A atoms must be rewritten from the
	// start of the file, so save to a unique sibling temp file and swap
	// atomically. Unique name so a concurrent tagger of the same directory
	// (import + stamp) never clobbers another writer's temp file.
	tmp, err := os.CreateTemp(filepath.Dir(filePath), filepath.Base(filePath)+".tagtmp-*")
	if err != nil {
		return fmt.Errorf("failed to create temp file for tagging: %w", err)
	}
	tmpPath := tmp.Name()
	// Cleans up on any error/panic exit; a no-op after a successful rename.
	defer os.Remove(tmpPath)

	saveErr := t.Save(tmp)
	closeErr := tmp.Close()
	if saveErr != nil {
		return fmt.Errorf("failed to save tags: %w", saveErr)
	}
	if closeErr != nil {
		return fmt.Errorf("failed to write tagged file: %w", closeErr)
	}
	// CreateTemp uses 0600; keep the original file's permissions so media
	// servers running as another user can still read the imported file.
	if info, statErr := os.Stat(filePath); statErr == nil {
		if chmodErr := os.Chmod(tmpPath, info.Mode().Perm()); chmodErr != nil {
			return fmt.Errorf("failed to preserve file permissions: %w", chmodErr)
		}
	}
	if err := os.Rename(tmpPath, filePath); err != nil {
		return fmt.Errorf("failed to replace file with tagged copy: %w", err)
	}
	return nil
}

// NormalizeAlbumTags stamps a canonical ALBUMARTIST onto a downloaded file so
// media servers group a multi-credit album under one artist instead of
// fragmenting it per-track (observed live: "Every Time I Die & Daryl Palumbo"
// folders splitting one album three ways). The monitored artist is the
// canonical album artist because acquisition items are keyed on it. When the
// file already carries an ALBUMARTIST tag it is left untouched.
// Best-effort: tagging errors are logged, never fail the import.
//
// Upstream note: audiometa v1.3.1 panicked on malformed/undecodable MP4 covr
// atoms; v3 returns errors instead. The recover() guard is kept as a backstop.
func (e *MetadataExtractor) NormalizeAlbumTags(filePath, albumArtist string) (err error) {
	if albumArtist == "" {
		return nil
	}
	ext := strings.ToLower(filepath.Ext(filePath))
	switch ext {
	case ".m4a", ".ogg":
		// A stamp is cosmetic — convert any panic into a clean skip and let
		// the caller's best-effort WARN handle it. (Backstop only; v3 returns
		// errors instead of panicking on malformed covr atoms.)
		defer func() {
			if r := recover(); r != nil {
				slog.Warn("audiometa panicked while stamping albumartist; skipping stamp",
					"path", filePath, "panic", r)
				err = nil
			}
		}()
		f, err := os.Open(filePath)
		if err != nil {
			return fmt.Errorf("open for albumartist stamp: %w", err)
		}
		defer f.Close()

		t, err := audiometa.OpenTag(f)
		if err != nil {
			return fmt.Errorf("open for albumartist stamp: %w", err)
		}
		if cur := t.GetAlbumArtist(); cur != "" {
			return nil
		}
		t.SetAlbumArtist(albumArtist)

		// v3 Save() writes to an io.Writer; M4A atoms must be rewritten from
		// the start of the file, so save to a unique sibling temp file and
		// swap atomically (unique name avoids clobbering a concurrent tagger).
		tmp, err := os.CreateTemp(filepath.Dir(filePath), filepath.Base(filePath)+".tagtmp-*")
		if err != nil {
			return fmt.Errorf("create temp for albumartist stamp: %w", err)
		}
		tmpPath := tmp.Name()
		// Cleans up on any error/panic exit; a no-op after a successful rename.
		defer os.Remove(tmpPath)

		saveErr := t.Save(tmp)
		closeErr := tmp.Close()
		if saveErr != nil {
			return fmt.Errorf("save albumartist stamp: %w", saveErr)
		}
		if closeErr != nil {
			return fmt.Errorf("write albumartist stamp: %w", closeErr)
		}
		// CreateTemp uses 0600; keep the original file's permissions.
		if info, statErr := os.Stat(filePath); statErr == nil {
			if chmodErr := os.Chmod(tmpPath, info.Mode().Perm()); chmodErr != nil {
				return fmt.Errorf("preserve file permissions: %w", chmodErr)
			}
		}
		if err := os.Rename(tmpPath, filePath); err != nil {
			return fmt.Errorf("replace file with stamped copy: %w", err)
		}
		return nil
	default:
		// MP3/FLAC: the underlying libs here do not expose ALBUMARTIST
		// writing cleanly; the canonical folder layout below still groups
		// the album for media servers.
		slog.Debug("albumartist stamp not supported for format", "ext", ext)
		return nil
	}
}

func (e *MetadataExtractor) embedMP3(filePath string, artData []byte) error {
	tag, err := id3v2.Open(filePath, id3v2.Options{Parse: true})
	if err != nil {
		return fmt.Errorf("failed to open mp3 for tagging: %w", err)
	}
	defer tag.Close()

	pic := id3v2.PictureFrame{
		Encoding:    id3v2.EncodingUTF8,
		MimeType:    detectImageMimeType(artData),
		PictureType: id3v2.PTFrontCover,
		Description: "Front Cover",
		Picture:     artData,
	}
	tag.AddAttachedPicture(pic)

	if err := tag.Save(); err != nil {
		return fmt.Errorf("failed to save mp3 tags: %w", err)
	}
	return nil
}

func (e *MetadataExtractor) embedFLAC(filePath string, artData []byte) error {
	f, err := flac.ParseFile(filePath)
	if err != nil {
		return fmt.Errorf("failed to parse flac: %w", err)
	}

	pic, err := flacpicture.NewFromImageData(
		flacpicture.PictureTypeFrontCover,
		"Front Cover",
		artData,
		detectImageMimeType(artData),
	)
	if err != nil {
		return fmt.Errorf("failed to create flac picture block: %w", err)
	}

	block := pic.Marshal()
	f.Meta = append(f.Meta, &block)
	if err := f.Save(filePath); err != nil {
		return fmt.Errorf("failed to save flac: %w", err)
	}
	return nil
}

func (e *MetadataExtractor) Extract(path string) (*AudioMetadata, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	m, err := tag.ReadFrom(f)
	if err != nil {
		return nil, fmt.Errorf("failed to read tags: %w", err)
	}

	metadata := &AudioMetadata{
		Artist:      m.Artist(),
		AlbumArtist: m.AlbumArtist(),
		Album:       m.Album(),
		Title:       m.Title(),
		Format:      string(m.FileType()),
	}

	track, _ := m.Track()
	metadata.TrackNumber = track
	metadata.Year = m.Year()

	info, err := os.Stat(path)
	if err == nil {
		metadata.FileSize = info.Size()
	}

	// Fallback for missing title
	if metadata.Title == "" {
		metadata.Title = strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
	}

	return metadata, nil
}

func (e *MetadataExtractor) IsAudioFile(path string) bool {
	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".mp3", ".flac", ".m4a", ".ogg", ".opus", ".wav":
		return true
	}
	return false
}

func (e *MetadataExtractor) SanitizeFilename(text string) string {
	if text == "" {
		return "Unknown"
	}

	// Replace problematic characters
	replacements := map[string]string{
		"/":  "-",
		"\\": "-",
		":":  "-",
		"*":  "",
		"?":  "",
		"\"": "'",
		"<":  "",
		">":  "",
		"|":  "-",
	}

	for old, new := range replacements {
		text = strings.ReplaceAll(text, old, new)
	}

	return strings.TrimSpace(text)
}

// GenerateLibraryPath builds the library path for an imported file. When
// metadata carries an AlbumArtist (canonical grouping artist) it is used for
// the folder so multi-credit tracks ("X & Y", guest features) all land in one
// album folder instead of fragmenting the album per track credit.
func (e *MetadataExtractor) GenerateLibraryPath(metadata *AudioMetadata, libraryRoot string) string {
	artist := metadata.AlbumArtist
	if artist == "" {
		artist = metadata.Artist
	}
	artist = e.SanitizeFilename(artist)
	album := e.SanitizeFilename(metadata.Album)
	if album == "" {
		album = "Unknown Album"
	}
	title := e.SanitizeFilename(metadata.Title)

	var filename string
	if metadata.TrackNumber > 0 {
		filename = fmt.Sprintf("%02d - %s%s", metadata.TrackNumber, title, e.getExt(metadata.Format))
	} else {
		filename = fmt.Sprintf("%s%s", title, e.getExt(metadata.Format))
	}

	return filepath.Join(libraryRoot, artist, album, filename)
}

func (e *MetadataExtractor) HashFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := md5.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

func (e *MetadataExtractor) Fingerprint(path string) (string, int, error) {
	// SECURITY: Validate path exists and is clean before passing to fpcalc
	cleanPath := filepath.Clean(path)
	if _, err := os.Stat(cleanPath); os.IsNotExist(err) {
		return "", 0, fmt.Errorf("file does not exist: %s", path)
	}

	cmd := exec.Command("fpcalc", "-json", "--", cleanPath)
	out, err := cmd.Output()
	if err != nil {
		return "", 0, fmt.Errorf("fpcalc failed: %w", err)
	}

	var result struct {
		Duration    int    `json:"duration"`
		Fingerprint string `json:"fingerprint"`
	}

	if err := json.Unmarshal(out, &result); err != nil {
		return "", 0, fmt.Errorf("failed to parse fpcalc output: %w", err)
	}

	return result.Fingerprint, result.Duration, nil
}

func (e *MetadataExtractor) getExt(format string) string {
	switch strings.ToUpper(format) {
	case "MP3":
		return ".mp3"
	case "FLAC":
		return ".flac"
	case "M4A", "AAC":
		return ".m4a"
	case "OGG":
		return ".ogg"
	case "OPUS":
		return ".opus"
	default:
		return ""
	}
}
